protocols/dcutr/src/behaviour.rs

```diff
@@ -0,0 +1,401 @@
+// Copyright 2021 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! [`NetworkBehaviour`] to act as a direct connection upgrade through relay node.
+
+use crate::handler;
+use crate::protocol;
+use either::Either;
+use libp2p_core::connection::{ConnectedPoint, ConnectionId};
+use libp2p_core::multiaddr::Protocol;
+use libp2p_core::{Multiaddr, PeerId};
+use libp2p_swarm::dial_opts::{self, DialOpts};
+use libp2p_swarm::{
+    DialError, IntoProtocolsHandler, NetworkBehaviour, NetworkBehaviourAction, NotifyHandler,
+    PollParameters, ProtocolsHandler, ProtocolsHandlerUpgrErr,
+};
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::task::{Context, Poll};
+use thiserror::Error;
+
+/// The events produced by the [`Behaviour`].
+#[derive(Debug)]
+pub enum Event {
+    InitiateDirectConnectionUpgrade {
```

Comment:
The naming of this variant doesn't indicate an event but a command. Is that on purpose?

Reply:
Unless I misunderstand the logic, this could be named Initiate**d**DirectConnectionUpgrade?

Reply:
`DirectConnectedUpgradeInitiated` might be even better but then it is not in line with `RemoteInitiatedDirectConnectionUpgrade`.

Reply:
Good idea. Renamed to `InitiatedDirectConnectionUpgrade ` for now with 8384b4e48b27e807a2b066cf21f9aea26b8ac1f7.

I am fine with `DirectConnectedUpgradeInitiated` as well.